nautilus_core/execution/src/matching_core/handlers.rs

```diff
@@ -0,0 +1,100 @@
+// -------------------------------------------------------------------------------------------------
+//  Copyright (C) 2015-2025 Nautech Systems Pty Ltd. All rights reserved.
+//  https://nautechsystems.io
+//
+//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
+//  You may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// -------------------------------------------------------------------------------------------------
+
+use std::{cell::RefCell, rc::Rc};
+
+use nautilus_model::orders::OrderAny;
+
+use crate::{
+    matching_engine::engine::OrderMatchingEngine, order_emulator::emulator::OrderEmulator,
+};
+
+pub trait FillMarketOrderHandler {
+    fn fill_market_order(&mut self, order: &OrderAny);
+}
+
+#[derive(Clone)]
+pub enum FillMarketOrderHandlerAny {
+    OrderMatchingEngine(Rc<RefCell<OrderMatchingEngine>>),
+    OrderEmulator(Rc<RefCell<OrderEmulator>>),
+}
+
+impl FillMarketOrderHandler for FillMarketOrderHandlerAny {
+    fn fill_market_order(&mut self, order: &OrderAny) {
+        match self {
+            FillMarketOrderHandlerAny::OrderMatchingEngine(engine) => {
+                engine.borrow_mut().fill_market_order(&mut order.clone());
+            }
+            FillMarketOrderHandlerAny::OrderEmulator(emulator) => {
+                emulator.borrow_mut().fill_market_order(&mut order.clone());
+            }
+        }
+    }
+}
```

Comment:
Depending on this the FillMarketOrderHandler is consumed it might be better to implement the `FillMarketOrderHandler` trait for the OrderMatchingEngine and OrderEmulator individually and store it as as Box<dyn FillMarketOrderHandler> or Rc<dyn FillMarketOrderHandler> that will reduce the number of enums and wrapper types created for this functionality.